Cloud Hypervisor support

hertrste opened the pull request:
This PR contains our most recent changes to add Cloud Hypervisor support to Nova.

hertrste committed:
libvirt/ch: add Cloud Hypervisor support

On-behalf-of: SAP philipp.schuster@sap.com
On-behalf-of: SAP stefan.kober@sap.com

Co-authored-by: Stefan Kober <stefan.hertrampf@cyberus-technology.de>
Co-authored-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
libvirt/ch: xxx fix rng device

In the old patchset from March, things used to work like this.
rng device probably not needed by SAP at the moment.

phip1611 committed:
libvirt/ch: apply SAP patches

xxx Patchset from christian

hertrste committed:
libvirt: fix readpty

Cloud Hypervisor uses a pty as a channel to and from the guest. To allow
openstack to display the console output of the VM, the pty must be read.

Reading from the pty has proven tricky, as it seems some input needs to
be provided and order to get output again. Therefore, we send some null
byte to the pty prior to reading. The reading is also done in a loop to
receive all available output.

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
libvirt/ch: allow NUMA tunings with CHV

Currently, the Nova Libvirt driver only enables numa tuning if QEMU/KVM
is the hypervisor type. As we have built NUMA support for the Libvirt
CHV driver, we now want to enable it in Nova too.

fwiesel commented on nova/virt/libvirt/driver.py:
```diff
@@ -8677,8 +8693,7 @@ def _has_numa_support(self):
         if (caps.host.cpu.arch in (fields.Architecture.I686,
                                    fields.Architecture.X86_64,
-                                   fields.Architecture.AARCH64) and
-                self._host.has_min_version(hv_type=host.HV_DRIVER_QEMU)):
```
I think, it would be better to test here for both `HV_DRIVER_QEMU` and `HV_DRIVER_CH`.
LXC and Xen do not support numa architecutres

hertrste replied:
There currently isn't a `HV_DRIVER_CH` variable defined. for qemu there is `HV_DRIVER_QEMU = "QEMU"`.

Following check does the comparison later on, where `hv_type ` is set to `HV_DRIVER_QEMU`:
```python
            if hv_type is not None:
                hypervisor_type = conn.getType()
                if hypervisor_type != hv_type:
                    return False
```

Do you have the possibility to check what `conn.getType()` returns in the case CHV is in use?

hertrste committed:
nix: support openstack-nix tests

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
ch: set host-passthrough as cpu mode per default

Nova requires some CPU model like host passthrough, host model, or
custom. Otherwise, things like NUMA tunings will not be configured
correctly later on. Therefore, we use host-passthrough per default for
cloud-hypervisor, which is the only mode currently supported.

hertrste committed:
test: use --wait in server delete

Otherwise the following check that the server really has been deleted
might fail, because the deleteion takes a while.

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
gitlab/ci: add openstack nixos test execution

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
nix: add CHV live migration test

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
nix: mv chv.nix to nix/

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
nix/tests: add NUMA tuning test

tpressure committed:
Merge branch 'gardenlinux-ci' into 'gardenlinux'

Add openstack-nix CI

See merge request cyberus/cloud/openstack-nova!6

hertrste committed:
nix/tests: use chvModule in all tests

hertrste committed:
ci: add hugepage numa test

tpressure committed:
Merge branch 'gardenlinux-ci-numa' into 'gardenlinux'

Add Hugepage support and Numa tunings

See merge request cyberus/cloud/openstack-nova!7

hertrste committed:
nix: nixpkgs update to 25.05

hertrste committed:
ch: add support for web serial console

In case CONF.serial_console.enabled is true, we now configure a TCP
based serial device for the Cloud Hypervisor domain. This allows access
to the serial console via the web based console.

hertrste committed:
libvirt: allow serial tcp output to be read

As with CHV we do not use the "console" tag but "serial". Therefore, add
that to the pattern matching used when trying to get serial/console
output.

hertrste committed:
nix: bump chv + libvirt

tpressure committed:
Merge branch 'web-serial-console' into 'gardenlinux'

Add support for web serial console

See merge request cyberus/cloud/openstack-nova!8

hertrste committed:
nix/test: allow usage of serial console

We do the necessary configuration for Nova to allow using the
interactive serial console via the horizon dashboard. We need to disable
VNC in order to make horizon use the serial console.

hertrste committed:
nix: bump openstack-nix

tpressure committed:
Merge branch 'web-serial-console' into 'gardenlinux'

Allow access to web serial console via dashboard

See merge request cyberus/cloud/openstack-nova!9

hertrste committed:
libvirt/ch: only set default cpu mode if none is set

Previously, we only supported "host-passthrough" as a cpu mode. Now that
we also support custom cpu models, we should only set the mode to
"host-passthrough" if no mode is explicitly specified via the nova
config.

On-behalf-of: SAP stefan.kober@sap.com

pkr4711 committed:
test: add end-to-end PCI addresses (BDF) tests

On-behalf-of: SAP paul.kroeher@sap.com

pkr4711 committed:
ci: add PCI addresses (BDF) tests to pipeline

On-behalf-of: SAP paul.kroeher@sap.com

pkr4711 committed:
test: add linter tools

On-behalf-of: SAP paul.kroeher@sap.com

pkr4711 committed:
docs: update documentation

On-behalf-of: SAP paul.kroeher@sap.com

pkr4711 committed:
Merge branch 'add-bdf-tests' into 'gardenlinux'

ci: add first end-to-end PCI addresses (BDF) tests

See merge request cyberus/cloud/openstack-nova!11

phip1611 committed:
nix: don't shadow pkgs of nixos modules

Passing `pkgs` as arguments in general has disadvantages as more
fine-grained parameters help to better keep track of actual
dependencies.

Further, we remove the shadowing of the function `pkgs` argument
and instead use `pkgs` from the NixOS module, as it is also
intended by Nix.

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
gitignore: add Nix artifacts

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
ci: remove deprecated alias

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

pkr4711 committed:
ci: add linter tools to pipeline

On-behalf-of: SAP paul.kroeher@sap.com

pkr4711 committed:
style: only formatting changes

On-behalf-of: SAP paul.kroeher@sap.com

hertrste committed:
Merge branch 'add-python-linter' into 'gardenlinux'

add linter tools

See merge request cyberus/cloud/openstack-nova!13

phip1611 committed:
nix/ci: format

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

hertrste committed:
ci: use randomized tmp directory

As our nixos tests are running on the same github runner sometimes, it
happens that there are clashes. As nix holds the QEMU disk images used
for the test VMs in /tmp and the name is deferred from the VMs name in
the nixos test, there can be clashes there.

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
ci: kill orphaned qemu processes

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
ci: execute test only in main pipeline

On-behalf-of: SAP stefan.kober@sap.com

hertrste committed:
Merge branch 'cpu-profiles-api-test' into 'gardenlinux'

Support CPU profile configuration

See merge request cyberus/cloud/openstack-nova!14

phip1611 committed:
ci: fix build job killing QEMU runs

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
ci/nix: accelerate CI

On-behalf-of: SAP philipp.schuster@sap.com
Signed-off-by: Philipp Schuster <philipp.schuster@cyberus-technology.de>

phip1611 committed:
Merge branch 'nix' into 'gardenlinux'

nix: various small cleanups + ci/nix speedup

See merge request cyberus/cloud/openstack-nova!12

pkr4711 committed:
fix: change to flake input of CHV

On-behalf-of: SAP paul.kroeher@sap.com

hertrste committed:
Merge branch 'use-flake-inputs' into 'gardenlinux'

fix: change to flake input of CHV

See merge request cyberus/cloud/openstack-nova!15

pkr4711 committed:
fix: change to flake input of libvirt

Signed-off-by: Paul Kroeher <paul.kroeher@cyberus-technology.de>

On-behalf-of: SAP paul.kroeher@sap.com

hertrste committed:
Merge branch 'use-flake' into 'gardenlinux'

fix: change to flake input of libvirt

See merge request cyberus/cloud/openstack-nova!16

hertrste commented:
Closing this as we have implemented a new development model in https://github.com/cobaltcore-dev/cobaltcore/issues/446
